src/Mvc/Mvc.ViewFeatures/src/RazorComponents/StaticComponentRenderer.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.RazorComponents
+{
+    internal class StaticComponentRenderer
+    {
+        private readonly HtmlEncoder _encoder;
+        private bool _initialized = false;
+
+        public StaticComponentRenderer(HtmlEncoder encoder)
+        {
+            _encoder = encoder;
+        }
+
+        public async Task<IEnumerable<string>> PrerenderComponentAsync(
+            ParameterCollection parameters,
+            HttpContext httpContext,
+            Type componentType)
+        {
+            var dispatcher = Renderer.CreateDefaultDispatcher();
+
+            InitializeUriHelper(httpContext);
+            using (var htmlRenderer = new HtmlRenderer(httpContext.RequestServices, _encoder.Encode, dispatcher))
+            {
+                var result = await dispatcher.InvokeAsync(() => htmlRenderer.RenderComponentAsync(
+                    componentType,
+                    parameters));
+                return result.Tokens;
+            }
+        }
+
+        private void InitializeUriHelper(HttpContext httpContext)
+        {
+            // The UriHelper might have been already initialized by a previous render.
+            if (!_initialized)
+            {
+                // This shouldn't be moved to the constructor as we want a request scoped service.
```

Comment:
I don't get this comment. Is this a singleton? Then it's not thread-safe. If it is anything but singleton we can DI a scoped service.

Reply:
Yeah, you are right. This should be a scoped service. The check is to not double initialize when you render multiple components with successive calls to RenderStaticComponentAsync()